File: Leetcode/Array/Python/three_sum.py
def three_sum(array):
    """Given an array S of n integers, are there elements a, b, c in S such that a + b + c = 0? 
    Find all unique triplets in the array which gives the sum of zero.

    Note:
        Elements in a triplet (a,b,c) must be in non-descending order. (ie, a ≤ b ≤ c)
        The solution set must not contain duplicate triplets.
        For example, given array S = {-1 0 1 2 -1 -4},
    A solution set is:
        (-1, 0, 1)
        (-1, -1, 2)
    """
    # sort array
    array.sort()

    # define vars
    solutions = set()
    unique_set = set(array)

    # check array length and other cases
    # if len(unique_set) == 1 and len(array) > 2 and 0 in unique_set:
    #     return [0, 0, 0]
    if len(array) < 2:
        return [-1]
    
    for i, v in enumerate(array[:-2]):
        j = i + 1
        k = len(array) - 1
        while j < k:
            current = v + array[j] + array[k]
            if current == 0:
                solutions.add(tuple([array[j], v, array[k]]))
                j += 1
                k -= 1
            elif current < 0:
                j += 1
            else:
                k -= 1
    return list(solutions)

            
three_sum(([-1, 0, 1, 2, -1, -4]))

File: Leetcode/Array/C++/threeSum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        int n = nums.size();
        if (n < 3) {
            return {};
        }
        vector<vector<int>> result;
        set<vector<int>> hash_set;

        sort(nums.begin(), nums.end());
        for (int i = 0; i < n - 2; i++) {
            int left = i + 1, right = n - 1;
            if (nums[i] > 0) {
                break;
            }
            if (i > 0 && nums[i] == nums[i - 1]) {
                continue;
            }
            while (left < right) {
                int sum = nums[i] + nums[left] + nums[right];
                if (sum == 0) {
                    vector<int> tmp(3);
                    tmp[0] = nums[i];
                    tmp[1] = nums[left];
                    tmp[2] = nums[right];
                    hash_set.insert(tmp);
                    left++;
                    right--;
                } 
                else if (sum < 0) {
                    left++;
                }
                else {
                    right--;
                }
            }
        }

        result = {hash_set.begin(), hash_set.end()};

        return result;
    }
};